plugins/inputs/ping_native/ping.go

```diff
@@ -0,0 +1,260 @@
+				return ip.String()
+			}
+		}
+	}
+
+	return ""
+}
+
+func (p *Ping) buildHostCache(acc telegraf.Accumulator) {
+	if p.hostCache != nil {
+		return
+	}
+
+	p.network = "ip4:icmp"
+	if runtime.GOOS == "darwin" || runtime.GOOS == "linux" {
+		p.network = "udp4"
+		if p.IPV6 {
+			p.network = "udp6"
+		}
+	} else if p.IPV6 {
+		p.network = "ip6:ipv6-icmp"
+	}
+
+	p.hostCache = []net.Addr{}
+	p.hosts = map[string]string{}
+
+	for _, host := range p.Hosts {
+		var addr net.Addr
+		var err error
+		if p.IPV6 {
+			addr, err = net.ResolveIPAddr("ip6", host)
+		} else {
+			addr, err = net.ResolveIPAddr("ip4", host)
+		}
+
+		switch runtime.GOOS {
+		case "js", "nacl", "plan9", "windows":
+			p.hosts[addr.String()] = host
+		default:
+			if p.IPV6 {
+				p.hosts["["+addr.String()+"]:0"] = host
+			} else {
+				p.hosts[addr.String()+":0"] = host
+			}
+		}
+
+		if a, ok := addr.(*net.IPAddr); ok && runtime.GOOS == "darwin" || runtime.GOOS == "linux" {
+			addr = &net.UDPAddr{IP: a.IP, Zone: a.Zone}
+		}
+
+		if err != nil {
+			acc.AddFields("ping", map[string]interface{}{"result_code": 1}, map[string]string{"source": host})
+			acc.AddError(err)
+			continue
+		}
+		if addr == nil {
+			continue
+		}
+		p.hostCache = append(p.hostCache, addr)
+	}
+}
+
+func (p *Ping) onRecv(pkt *ping.Packet) {
+	p.rcvdCache[pkt.IPAddr] = pingResults{ttl: pkt.TTL}
+}
+
+func (p *Ping) onFinish(acc telegraf.Accumulator) func(stats *ping.Statistics) {
+	return func(stats *ping.Statistics) {
+		min := float64(stats.MinRTT.Nanoseconds()) / float64(time.Millisecond)
+		avg := float64(stats.AvgRTT.Nanoseconds()) / float64(time.Millisecond)
+		max := float64(stats.MaxRTT.Nanoseconds()) / float64(time.Millisecond)
+		stddev := float64(stats.StdDevRTT.Nanoseconds()) / float64(time.Millisecond)
+
+		tags := map[string]string{"source": p.hosts[stats.Addr]}
+		fields := map[string]interface{}{
+			"result_code":         0,
+			"packets_transmitted": int(stats.PacketsSent),
+			"packets_received":    int(stats.PacketsRecv),
+			"percent_packet_loss": stats.PacketLoss,
+		}
+
+		if p.rcvdCache[stats.Addr].ttl > 0 {
+			fields["ttl"] = p.rcvdCache[stats.Addr].ttl
+		}
+		if min >= 0 {
+			fields["minimum_response_ms"] = min
+		}
+		if avg >= 0 {
+			fields["average_response_ms"] = avg
+		}
+		if max >= 0 {
+			fields["maximum_response_ms"] = max
+		}
+		if stddev >= 0 {
+			fields["standard_deviation_ms"] = stddev
+		}
+
+		acc.AddFields("ping", fields, tags)
+	}
+}
+
+// Gather gathers ping metrics via native ping implementation.
+func (p *Ping) Gather(acc telegraf.Accumulator) error {
+	if p.Interface != "" && p.listenAddr == "" {
+		p.listenAddr = getAddr(p.Interface)
+	}
+
+	p.buildHostCache(acc)
+	if len(p.hostCache) == 0 {
+		return errors.New("no valid hosts to ping")
+	}
+
+	conn, err := ping.Listen(p.network, p.listenAddr)
+	defer conn.Close()
+	if err != nil {
+		return err
+	}
+
+	if p.Count < 0 {
+		p.Count = 0
+	}
+
+	var size uint = 56
+	if p.IPV6 {
+		size = 128
+	}
```

Comment:
The setting of size is quite mysterious, I assume this is:
>       -s packetsize
>           Specifies the number of data bytes to be sent. The default is 56,
>           which translates into 64 ICMP data bytes when combined with the 8
>           bytes of ICMP header data.

Why is it increased for IPv6?  It seems like these should be the defaults for the ping library that we don't need to deal with here.